Configure ingressController to use routeSelector and configure external ingress service

dystewart opened the pull request.

dystewart committed:
Change ingressController to use routeSelector

To allow for more granularity, we are opting to go with a route selector as opposed to a namespace
selector for the ingress controller. The idea here is that we're filtering the traffic serviced by the ingress
controller at the route level as opposed to the entire namespace.

dystewart committed:
Configure address for external ingress service

This service is used with the external ingressController and connects with our public metallb loadBalancer
The important fields to pay attention to here are the:
metadata.annotations metallb.universe.tf/address-pool: public

spec.loadBalancerIP: 199.94.61.6
This is the ip where we are servicing external traffic

larsks reviewed (dismissed):
I've noted one additional change we need to make here.

Since you're going to be making changes *anyway*, I would suggest making the commit messages slightly more descriptive:

- For the "Change ingressController to use routeSelector", consider adding to the commit body our reasons for that change.

- For "Configure address for external ingress service", it's probably worth noting which fields in particular we're adding with this change (and why), since otherwise it's not obvious.

In either or both cases, consider adding a link to the relevant documentation if you think it's appropriate (see e.g. 480edd0058553419819257c6b01f8394a9bd2054 for an example of how I typically do that).

larsks commented on cluster-scope/overlays/nerc-ocp-prod/ingresscontrollers/external-apps-ingress-controller.yaml:
```diff
@@ -15,6 +15,6 @@ spec:
-  namespaceSelector:
+  routeSelector:
     matchLabels:
       type: external
```
Actually, here is one more change I think we should make. I think that `type` is probably too generic. Let's (a) pick a more descriptive name, and (b) let's namespace it like we have for other locally relevant labels.

How about:

```
routerSelector:
  matchLabels:
    nerc.mghpcc.org/external-ingress: "true"
```

This is the same label we use to identify nodes that host the external ingress address (and placement for the external ingress service).

dystewart replied:
@larsks good points, updated the commit messages and changed the label

larsks committed:
Rename external-apps-ingress-controller to external-apps

We already know it's an ingress controller.

larsks committed:
Patch external ingress service via argocd post-sync hook

When we create an IngressController, the IngressController creates a
Service. We need to modify that service resource to add both an annotation
and a loadBalancerIP, but ArgoCD doesn't allow us to patch existing
resources.

The best we can do is apply patches using an ArgoCD post-sync hook [1].

This commit adds a post-sync Job that takes care of patching the Service
resource. This job will run after every sync operation.

[1]: https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/

larsks committed:
Relocate external ingress related resources to feature/external-ingress

Keep all the external-ingress resources together so they're easier to
manage.

naved001 reviewed (approved):
looks good. I don't know a lot about patch operator, but is there a reason why we aren't using that?
